src/dns.rs

```diff
@@ -0,0 +1,305 @@
+
+    // Terminate the final string, replacing the last '.'
+    parsed_name[name_len - 1] = '\0' as u8;
+    // Return pointer to first char after the name
+    return unsafe { label.offset(1) };
+}
+
+fn parse_dns_request(
+    req: &mut [u8],
+    req_len: usize,
+    dns_reply: &mut [u8],
+    dns_reply_max_len: usize,
+) -> Option<usize> {
+    if req_len > dns_reply_max_len {
+        return None;
+    }
+
+    // Prepare the reply
+    dns_reply.fill(0);
+    (&mut dns_reply[0..req_len]).copy_from_slice(&req[0..req_len]);
+
+    let header_len = mem::size_of::<DnsHeader>();
+    let (header_bytes, rest) = dns_reply.split_at_mut(header_len);
+
+    // Endianess of NW packet different from chip
+    let header = unsafe {
+        header_bytes
+            .as_mut_ptr()
+            .cast::<DnsHeader>()
+            .as_mut()
+            .unwrap()
+    };
+
+    debug!(
+        "DNS query with header id: 0x{:X}, flags: 0x{:X}, qd_count: {}",
+        ntohs(header.id),
+        ntohs(header.flags),
+        ntohs(header.qd_count)
+    );
+
+    // Not a standard query
+    if (header.flags & OPCODE_MASK) != 0 {
+        return None;
+    }
+
+    // Set question response flag
+    header.flags |= QR_FLAG;
+
+    let qd_count = ntohs(header.qd_count);
+    header.an_count = htons(qd_count);
+
+    let reply_len = qd_count as usize * mem::size_of::<DnsAnswer>() + req_len;
+    if reply_len > dns_reply_max_len {
+        return None;
+    }
+
+    // Pointer to current answer and question
+    let (questions, answers) = rest.split_at_mut(req_len - header_len);
+    let cur_qd_ptr = questions.as_mut_ptr();
+    let mut cur_ans_ptr = answers.as_mut_ptr();
+    let mut name: [u8; 128] = [0; 128];
+
+    // Respond to all questions with the ESP32's IP address
+    for i in 0..qd_count {
+        debug!("answering question {}", i);
+        let name_end_ptr = parse_dns_name(cur_qd_ptr, &mut name);
+        if name_end_ptr.is_null() {
+            error!("failed to parse DNS question: {:?}", unsafe {
+                CStr::from_ptr(cur_qd_ptr as _)
+            });
+            return None;
+        }
+
+        let question = unsafe { name_end_ptr.cast::<DnsQuestion>().as_mut().unwrap() };
+        let qd_type = ntohs(question.typ);
+        let qd_class = ntohs(question.class);
+
+        info!(
+            "received type: {} | class: {} | question for: {:?}",
+            qd_type,
+            qd_class,
+            unsafe { CStr::from_ptr(name.as_ptr() as _) }
+        );
+
+        if qd_type == QD_TYPE_A {
+            let answer = unsafe { cur_ans_ptr.cast::<DnsAnswer>().as_mut().unwrap() };
+
+            let ptr_offset = unsafe { cur_qd_ptr.offset_from(dns_reply.as_ptr()) };
+            answer.name_ptr = htons((0xC000 | ptr_offset) as u16);
+            answer.typ = htons(qd_type);
+            answer.class = htons(qd_class);
+            answer.ttl = htonl(ANS_TTL_SEC);
+
+            let mut ip_info = esp_netif_ip_info_t::default();
```

Comment:
In an implementation which is independent of ESP-IDF, we would want the IP address to be abstracted away, via a trait, or passed at construction/start time.